oak/server/rust/oak_runtime/src/node/grpc_server.rs

```diff
@@ -0,0 +1,275 @@
+            Ok(body) => body,
+            Err(error) => {
+                // TODO: error? can module return an Err based on the client request?
+                error!("Couldn't process gRPC response: {}", error);
+                return Ok(Self::http_response(
+                    http::StatusCode::INTERNAL_SERVER_ERROR,
+                    vec![],
+                ));
+            }
+        };
+
+        // Send response back to the gRPC client.
+        Ok(Self::http_response(http::StatusCode::OK, response_body))
+    }
+    
+    fn decode_grpc_request(grpc_method: &str, http_body: &dyn hyper::body::Buf) -> Result<Box<GrpcRequest>, String> {
+        // Parse HTTP body as into a Protobuf message.
+        match protobuf::parse_from_bytes::<Any>(http_body.bytes()) {
+            Ok(protobuf_body) => {
+                // Create a gRPC request.
+                match oak::grpc::encap_request(&protobuf_body, None, grpc_method) {
+                    Some(grpc_request) => Ok(Box::new(grpc_request)),
+                    None => Err("Failed to parse Protobuf message".to_string()),
+                }
+            },
+            Err(protobuf_error) => {
+                let error = format!("Failed to build GrpcRequest {}", protobuf_error);
+                Err(error)
+            },
+        }
+    }
+
+    // Creates an HTTP response message.
+    fn http_response(status: http::StatusCode, body: Vec<u8>) -> hyper::Response<hyper::Body> {
+        let mut response = hyper::Response::new(hyper::Body::from(body));
+        *response.status_mut() = status;
+        response
+    }
+
+    /// Processes a gRPC request from the client and forwards it to the Oak node.
+    /// Returns a channel handle for reading the response from the Oak node or an error. 
+    fn process_request(&self, request: Box<GrpcRequest>) -> Result<Handle, String> {
+        // Create a pair of temporary channels to pass the request to the Oak node and to receive the response.
+        let (request_writer, request_reader) = self.runtime.new_channel(
+            self.node_id,
+            &oak_abi::label::Label::public_trusted()
+        );
+        let (response_writer, response_reader) = self.runtime.new_channel(
+            self.node_id,
+            &oak_abi::label::Label::public_trusted()
+        );
+        
+        // Create a notification message and attach the method-invocation specific channels to it.
+        let notification = crate::Message {
+            data: vec![],
+            channels: vec![request_reader, response_writer],
+        };
+        
+        // Serialize gRPC request into a message.
+        let message = crate::Message { data: vec![], channels: vec![] };
+        if let Err(protobuf_error) = request.write_to_writer(&mut message.data) {
+            let error = format!("Failed to serialize GrpcRequest message: {}", protobuf_error);
+            return Err(error);
+        }
```

Comment:
It would be nice to have more idiomatic error handling using `?`, it would make things more readable.

Reply:
Done